### PR TITLE
issue 466.1: initial commit of FlushFeature redesign with WAL markers

### PR DESCRIPTION
**Pull request by vasiliy-arangodb**

### CODE REVIEW

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2761/

File: arangod/MMFiles/MMFilesWalRecoverState.cpp [hunk 1]
```diff
@@ -444,6 +444,31 @@ bool MMFilesWalRecoverState::InitialScanMarker(MMFilesMarker const* marker, void
 bool MMFilesWalRecoverState::ReplayMarker(MMFilesMarker const* marker,
                                           void* data, MMFilesDatafile* datafile) {
   MMFilesWalRecoverState* state = reinterpret_cast<MMFilesWalRecoverState*>(data);
+  auto visitHelpers = std::shared_ptr<MMFilesWalRecoverState>(
+    state,
+    [marker](MMFilesWalRecoverState* state)->void {
```
Comment (gnusi) [inline arangod/MMFiles/MMFilesWalRecoverState.cpp]: destructor is not marked as `noexcept` which may cause the whole process termination

File: arangod/RestServer/FlushFeature.cpp [hunk 2]
```diff
@@ -20,20 +20,421 @@
...
+
+  if (!typeSlice.isString()) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      "failed to find a string 'type' attribute inside the supplied marker body while applying 'Flush' recovery marker"
+    );
+  }
+
+  auto type = typeSlice.copyString();
+  auto& callbacks = getFlushRecoveryCallbacks();
```
Comment (gnusi) [inline arangod/RestServer/FlushFeature.cpp]: why don't we synchronize access to static variable?
if that's intended, please leave a comment that this particular function is designed to be used in single-threaded environment only

File: arangod/RestServer/FlushFeature.cpp [hunk 3]
```diff
@@ -59,6 +460,123 @@ void FlushFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
                      arangodb::options::makeFlags(arangodb::options::Flags::Hidden));
 }
 
+/*static*/ bool FlushFeature::registerFlushRecoveryCallback(
+    std::string const& type,
+    FlushRecoveryCallback const& callback
+) {
+  return !callback // skip empty callbacks
+         || getFlushRecoveryCallbacks().emplace(type, callback).second;
```
Comment (gnusi) [inline arangod/RestServer/FlushFeature.cpp]: why don't we synchronize access to static variable?
if that's intended, please leave a comment that this particular function is designed to be used in single-threaded environment only

File: arangod/RestServer/FlushFeature.h [hunk 4]
```diff
@@ -68,8 +112,10 @@ class FlushFeature final : public application_features::ApplicationFeature {
 
   basics::ReadWriteLock _callbacksLock;
   std::unordered_map<void*, FlushCallback> _callbacks;
+  std::unordered_set<std::shared_ptr<FlushSubscriptionBase>> _flushSubscriptions;
+  std::mutex _flushSubscriptionsMutex;
```
Comment (gnusi) [inline arangod/RestServer/FlushFeature.h]: please rename `_flushSubscriptionsMutex` in to `_flushSubscriptionsLock`

File: arangod/Aql/Ast.cpp [hunk 5]
```diff
@@ -2073,6 +2073,38 @@ void Ast::validateAndOptimize() {
   _root = traverseAndModify(_root, preVisitor, visitor, postVisitor);
 }
 
+/// @brief determines the variables referenced in an expression
```
Comment (jsteemann) [inline arangod/Aql/Ast.cpp]: I don't see any need for this method.
We have the same method that takes an arangodb::HashSet as its second parameter. Can't we just use that?

File: arangod/Aql/Ast.h [hunk 6]
```diff
@@ -416,6 +416,7 @@ class Ast {
   void validateAndOptimize();
 
   /// @brief determines the variables referenced in an expression
+  static void getReferencedVariables(AstNode const*, std::unordered_set<Variable const*>&);
```
Comment (jsteemann) [inline arangod/Aql/Ast.h]: I don't see a value in this given that the other method already exists.

File: arangod/RestServer/FlushFeature.cpp [hunk 7]
```diff
@@ -20,20 +20,421 @@
...
+  MMFilesFlushMarker(
+      TRI_voc_tick_t databaseId,
+      arangodb::velocypack::Slice const& slice
+  ): _databaseId(databaseId), _slice(slice) {
+  }
+
+  TRI_voc_tick_t databaseId() const noexcept { return _databaseId; }
+
+  // always 0 since not yet in any datafile
+  TRI_voc_fid_t fid() const override final { return 0; }
```
Comment (jsteemann) [inline arangod/RestServer/FlushFeature.cpp]: why is this needed?

File: arangod/Aql/Ast.cpp [hunk 8]
```diff
@@ -2073,6 +2073,38 @@ void Ast::validateAndOptimize() {
...
+void Ast::getReferencedVariables(
```
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.cpp]: without this there is a build failure
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/3959/

File: arangod/MMFiles/MMFilesWalRecoverState.cpp [hunk 9]
```diff
@@ -444,6 +444,31 @@ bool MMFilesWalRecoverState::InitialScanMarker(MMFilesMarker const* marker, void
 bool MMFilesWalRecoverState::ReplayMarker(MMFilesMarker const* marker,
                                           void* data, MMFilesDatafile* datafile) {
   MMFilesWalRecoverState* state = reinterpret_cast<MMFilesWalRecoverState*>(data);
+  auto visitHelpers = std::shared_ptr<MMFilesWalRecoverState>(
```
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/MMFiles/MMFilesWalRecoverState.cpp]: a shared pointer is used so that its distructor is be called

File: arangod/MMFiles/MMFilesWalRecoverState.cpp (hunk 1 above)
Comment (vasiliy-arangodb) [reply arangod/MMFiles/MMFilesWalRecoverState.cpp]: true

File: arangod/RestServer/FlushFeature.cpp (hunk 2 above)
Comment (vasiliy-arangodb) [reply arangod/RestServer/FlushFeature.cpp]: please see above

File: arangod/RestServer/FlushFeature.cpp (hunk 3 above)
Comment (vasiliy-arangodb) [reply arangod/RestServer/FlushFeature.cpp]: please see above

File: arangod/RestServer/FlushFeature.h (hunk 4 above)
Comment (vasiliy-arangodb) [reply arangod/RestServer/FlushFeature.h]: it's only a mutex, not a scoped lock though

File: arangod/MMFiles/MMFilesEngine.cpp [hunk 10]
```diff
@@ -379,6 +385,36 @@ Result MMFilesEngine::persistLocalDocumentIds(TRI_vocbase_t& vocbase) {
...
+  try {
+    getRecoveryHelpers().emplace_back(&helper);
+  } catch (std::bad_alloc const&) {
+    return arangodb::Result(TRI_ERROR_OUT_OF_MEMORY);
+  }
+
+  return arangodb::Result();
+}
+
+/*static*/ bool MMFilesEngine::visitRecoveryHelpers(
```
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/MMFiles/MMFilesEngine.cpp]: see above

File: arangod/MMFiles/MMFilesEngine.cpp [hunk 11]
```diff
@@ -379,6 +385,36 @@ Result MMFilesEngine::persistLocalDocumentIds(TRI_vocbase_t& vocbase) {
   return result;
 }
 
+/*static*/ arangodb::Result MMFilesEngine::registerRecoveryHelper(
+  MMFilesRecoveryHelper const& helper
+) {
+  try {
+    getRecoveryHelpers().emplace_back(&helper);
```
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/MMFiles/MMFilesEngine.cpp]: done

File: arangod/Aql/Ast.cpp (hunk 5 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.cpp]: without this there is a build failure http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/3959/

File: arangod/Aql/Ast.h (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.h]: without this there is a build failure http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/3959/

File: arangod/RestServer/FlushFeature.cpp (hunk 7 above)
Comment (vasiliy-arangodb) [reply arangod/RestServer/FlushFeature.cpp]: databaseId() because _databaseId member is private
fid() because MMFilesWalMarker::fid() is abstract

File: arangod/MMFiles/MMFilesWalRecoverState.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/MMFiles/MMFilesWalRecoverState.cpp]: you can use `TRI_DEFER(...)` to ensure that function will be called eventually

File: arangod/Aql/Ast.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/Aql/Ast.cpp]: you have to merge the recent `devel` into your branch then

File: arangod/Aql/Ast.h (hunk 6 above)
Comment (gnusi) [reply arangod/Aql/Ast.h]: you have to merge the recent `devel` into your branch then

File: arangod/Aql/Ast.cpp (hunk 5 above)
Comment (gnusi) [reply arangod/Aql/Ast.cpp]: you have to merge the recent devel into your branch then

File: arangod/Aql/Ast.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.cpp]: that branch was made from recent devel, and this is a build failure in community, not enterprise

File: arangod/Aql/Ast.h (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.h]: that branch was made from recent devel, and this is a build failure in community, not enterprise

File: arangod/Aql/Ast.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/Aql/Ast.cpp]: *current* `devel` has only 1 version of `getReferencedVariables` function and compiles without any errors: https://github.com/arangodb/arangodb/blob/devel/arangod/Aql/Ast.h
Probably you have to pull the recent `devel` into your branch once again.

File: arangod/Aql/Ast.h (hunk 6 above)
Comment (gnusi) [reply arangod/Aql/Ast.h]: current devel has only 1 version of getReferencedVariables function and compiles without any errors: https://github.com/arangodb/arangodb/blob/devel/arangod/Aql/Ast.h
Probably you have to pull the recent devel into your branch once again.

File: arangod/Aql/Ast.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.cpp]: the branch was based on a clean devel clone as of this morning

File: arangod/Aql/Ast.h (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/Aql/Ast.h]: the branch was based on a clean devel clone as of this morning, as per the git history

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/2770/

**Comment by vasiliy-arangodb** [comment]
replaced by https://github.com/arangodb/arangodb/pull/7980
